SimulationAgent/Agent.cs

```diff
@@ -89,6 +93,68 @@ public void Stop()
             this.runner.Stop();
         }
 
+        public async Task AddDevice(string deviceId, string modelId)
+        {
+            // currently only one simulation is supported
+            var simulation = (await this.simulations.GetListAsync()).FirstOrDefault();
+
+            if (simulation != null && this.IsDeviceModelIdValidAsync(modelId))
+            {
+                try
+                {
+                    await this.AddDeviceToSimulationRecordAsync(simulation, deviceId, modelId);
+
+                    if (this.running)
+                    {
+                        this.log.Info("Add device to running simulation");
+                        await this.runner.AddDevice(deviceId, modelId);
+                    }
+                    else
+                    {
+                        this.log.Info("Add device to IoT Hub");
+                        await this.simulations.AddDeviceAsync(deviceId);
+                    }
+                }
+                catch (Exception e)
+                {
+                    this.log.Debug("Error while adding new device", () => new { e });
+                    throw new Exception("Error while adding a new device");
+                }
+            }
+        }
+
+        public async Task DeleteDevices(List<string> ids)
+        {
+            this.log.Info("Update simulation");
+
+            // currently only one simulation is supported
+            var simulation = (await this.simulations.GetListAsync()).FirstOrDefault();
```

Comment:
why not use `this.simulation` ?  also, add `this.simulation = null` in `Stop()`

Reply:
good point, updated both.